### ISSUE TITLE
Is it thread safe to notify blocked thread during release when fair = true

### ISSUE THREAD

**Issue by 0ctopus13prime** [anandkulkarnisg/Semaphore#1]
Hello, when I looking through your source code it noticed unsafe access to `m_queue` without lock guarding

https://github.com/anandkulkarnisg/Semaphore/blob/master/Semaphore.cpp#L251

```
	// This is actually interesting. If the mode is fair then we signal only if we know that the permits that we have released will serve the first item in the queue.
	// We also need to signal in fair mode as long as wait queue exists and we can serve the front item currently. This fixes a rare bug where notify may miss out and some acquire may keep waiting.
	if(isFair())
	{
		while(hasQueuedThreads() && get<1>(*m_queue.begin()) <=m_permits)  <--------- Here!
		{
			m_cond.notify_all();
		}
	}
```

As you can see, `*m_queue.begin()` is performed without proper guarding.
Because `exclusiveLock` above the if-statement was unlocked already, during accessing element in `begin()` another thread might barging in and could do perform `push_back` 
https://github.com/anandkulkarnisg/Semaphore/blob/master/Semaphore.cpp#L88

Just for sure, is it OK to do it?
Please tell me if I misunderstood.

Thanks in advance

**Comment by anandkulkarnisg** [comment]
Your question is valid , but remember two things.

a. If i keep lock and signal the other threads that are awaiting for permits or are queuing for it [ because mode is fair ] then none of them will be able to get the lock and do anything useful. so its fine that the lock is unlocked first and then signaled so that other threads can pick the lock and try to take the permits. only one of them will succeed at a time.

b. m_queue.begin() may not change since push_back will add to the tail or end of the queue. I agree that if there is only single element this may cause issues...so i need to look into it further. If you dont mind can you reproduce edge case where you think it might blow up. I can fix it.

**Comment by anandkulkarnisg** [comment]
@0ctopus13prime , I might have solved the bug. I will test it extensively and release the fix. Its a patch to acquire Internal method.
Also the while loop of sending signal over m_cond [ which is bit ugly and inefficient ] is resolved. not needed anymore. Expect a patch/solution by end of this week. ETA : 25 to 27/09/2020.

**Comment by anandkulkarnisg** [comment]
@0ctopus13prime  , Bug fix is done with commit : 99c6235ed53bc73bf8b949fe21fbb12eca5487f3 , Please pull. I have tested the fix and seems to work without any deadlocks by using run.sh 100 command [ or any number of iterations you desire ] . I have tested it upto 500 iterations of different simulated runs.
